### DISCUSSION TITLE
Firestore Listener

### DISCUSSION THREAD

**Discussion by itsfintom** [SchneiderInfosystems/FB4D#117] [Q&A] [answered]
Have a question to how the Listener should normally work:

Adding targets to the listener with SubscribeQuery
Then start the listener.
After starting I get an initial document back, even the document has not been changed in the cloud
(is this intended to be so ?, Always get an initial response from all target collections / documents)
Now change a document / documents in the cloud..All fine changes are reported..
Now, stop the listener.
Then start the listener again, without adding new targets.
**No changes are detected anymore.**
In the code of TFirestoreDatabase.StopListener, the fListener is Free-d.
The fTargets is a private variable in TFSListenerThread..So it gets also flushed.

In the Wiki the behaviour is described like that:
Add Targets
Start Listener
If we need to unsubscribe or subscribe new targets, we need to Stop the Listener
After changing the subscriptions, just Start the Listener again.

Apparently this does not work as expected.

Am I getting something wrong here, or can we consider this as a bug ?

I tried it on iOS

**Comment by SchneiderInfosystems** [comment, marked as answer]
The Firestore always sends the whole document when creating a listener. FB4D only forwards this message. But I find this solution quite practical because in most cases you would need a get before. 

Yes, after StopListener you must call SubscribeDocument again. Another approach would require an unsubscribeDocument. For me, this solution was simpler to understand.

I cannot reproduce your reported defect, why you don't get changed documents after the second StartListener. Please try also the DemoFB4D. If you are sure that this problem only appears under iOS, can you show me a logfile of a run under iOS? Does an exception occur?

It seems important to me to know that the rewriting of a document only does not necessarily lead to a new message. The content of the document must be changed, otherwise, you will not receive a new message.

Is this information clear enough to help?

**Reply by itsfintom** [reply]
Thank you for your answer.

Just to confirm:

So, lets say I subscribe 3 collections, then because of a user added a collection which I would listen to as well, I need to:
Stop the listener
Add my previous 3 collections again,
Then add the new one.. (total 4)
And then start the Listener again..
If this is the case, the unsubscribe functions are useless..??

From the WIKI text, it sounds like it keeps the 3..So after stopping the listener, I only have to add the new ones, and start again.

In my real world application I do have a single root collection.
Documents in this collection contain sub collections.
Now I have a listener to the root collection, and to each individuals document sub collection.
If a user adds a document in the root collection, I need to add a new listener to that documents sub collection.


Thank you

**Reply by SchneiderInfosystems** [reply]
After deeper investigation, I agree with you completely. The interface was originally designed to offer a manual unsubsribe. I didn't remember that yesterday when I sent the last reply. In fact, after stop listener it needs a new listener, because a thread cannot be restartet again. 
I can bring now a fix that will save previously installed targets over to the new listener. Therefore, I would introduce a new flag (RemoveAllSubscription: boolean = true) in the StopListener function, which controls this action, to ensure backward compatibility.

**Reply by SchneiderInfosystems** [reply]
Please check my last commit with the changes. The current solution prevents also to register a target more than once. That is why the flag RemoveAllSubscription is not that relevant because a second subscribe of the same does not result in twice registration.
Can you carefully test the solution?

**Reply by itsfintom** [reply]
Thank you
This is what I have implemented myself yesterday, I used an "external" TList object to hold a collection/document collective string, and only update this one. If the list content has changed, this is checked in an idle tread, it will stop the listener, subscribe new based on my List, and the Start the listener again. It works.
One more thing I have noticed is, that the stop listener must not be called in the context of the listener thread. This would be the case, if a document change is reported by the listener, and when handling this call, one tries to stop the listener. This will kill the app on an iOs. Correct way is to set a flag, and let another thread (main) to the stop and restart. This is what i do with the List object....
Maybe you need to dress this as well

Will test the commit, and let you know

**Reply by itsfintom** [reply]
Have tested the stuffs, but I guess there is another issue. Still tracking it, but to let you know..
When starting and stopping listeners, it can happen that a document is currently processing, and handled in the OnFSChangeDoc Event handler. Wether it is synchronised or not, in that scenario the APP (iOS in this case) will crash and close. I have tracked it down to that point so far. To get around the issue I have implemented a state machine with the following states (STOPED,STOPING,RUN,DOC)..I run a simple timer (for the sake of test only) where I check if there is a new listener to add, and based on the state machine, I stop the listener (only when in RUN), set the state to STOPING..In the Event handler (yours) set it to STOPED when finally stoped. When in the DOC processing, I set it to DOC, to prevent any actions.. It works so far now, but please investigate that. In principle, the stop has to wait until any actions triggered by the thread are finished, and no new ones are started.. Will test a bit more to track it down, but don't want to mess with your code, because you simply know the internals better...
Regards Thomas

**Reply by SchneiderInfosystems** [reply]
From my point of view, your database structure is wrong, because you often have to start/stop the listener during operation. Can't you work with one collection monitor the whole collection for all changed documents within this collection? Please check my `FirestoreSimpleChat` application. Here every user can create their own document within the collection `Chat` and all other users have a running listener to this collection to be informed when a new document was written.
And clearly, the listener cannot be restarted from a callback. This must be done from the GUI thread. You usually start a new thread in the `onTerminate` method. So here a SuperThread would be necessary, which monitors everything and restarts the thread. But as said it is probably easier to keep the listener running all the time.

**Reply by itsfintom** [reply]
The listener is restarted from the GUI thread, and not a callback, that is clear. I use an event mechanism to trigger it. Still the scenario above holds true. If in any moment you stop a listener, but it is currently not "idle" it will crash. The actual root cause is unclear, but for that I would need to dig deeper into your code.

And it is not done frequent, just maybe once every 10 minutes...Only if somebody adds a new document into a collection, which itself holds a sub collection to be monitored.

The database structure uses documents with sub collection. 
Actually if it were possible to query on a CollectionGroup it would solve that issue.
The collections to monitor (sub collections) have always the same name "pages".. So I could simply index them and run a CollectionGroup query.
I could not figure out how to manage that with FB4D...
Have tried the collections,desdendent =true...But that's another animal I guess

Regards
Thomas

**Reply by itsfintom** [reply]
Have found FB4D supports the structuredQuery..."allDescendants":true....So CollectionGroup Query is possible...I will change my code to use it. Just ran a simple test and it works like expected. Somehow when I played with the descendant=true in .CreateForCollection in the first place, it was not working, probably because of rules were set wrong in firebase...Or a typo.. 
This is the better approach and avoids the listener changes.

Thank you
Thomas

**Reply by SchneiderInfosystems** [reply]
Updated: Sorry for my too fast reply yesterday, I have now investigated the situation in more detail. IMHO `IncludesDescendants` is only required to notify for changes in sub-documents of a collection. If "normal" documents are changed in a collection this flag can be false.
I have now added such a checkbox in the Inro Demo for Firestore listener when using collections.
